Mathlib/GameTheory/FairDivision/Discrete/Basic.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright (c) 2024 Shreyas Srinivas. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Shreyas Srinivas
+-/
+
+import Mathlib.Algebra.BigOperators.Group.Finset
+import Mathlib.Data.Real.Basic
+import Mathlib.Tactic.Linarith
+import Mathlib.Tactic.Ring.Basic
+
+
+namespace Mathlib
+
+namespace DiscreteFairDivision
+
+variable {Agents Goods : Type}
+variable [Fintype Agents] [Fintype Goods]
+variable [DecidableEq Agents]
+variable [DecidableEq Goods] -- Indivisible goods only
+
+
+
+structure Alloc where
+  assign : Agents → Finset Goods
+  disjoint : ∀ a b : Agents, a ≠ b → assign a ∩ assign b = ∅
+
+structure Val where
+  val : Agents → Finset Goods → ℝ
+  empty : ∀ a : Agents, val a ∅ = 0
+  pos : ∀ a : Agents, ∀ s : Finset Goods, val a s ≥ 0
+  mono : ∀ a : Agents, ∀ s t : Finset Goods,
+    s ⊆ t → val a s ≤ val a t
+
+
+variable (alloc : @Alloc Agents Goods _)
+variable (v : @Val Agents Goods)
+
+def valSingle (a : Agents) (g : Goods) :=
+  v.val a {g}
+
+def isEmpty (a : Agents):=
+  alloc.assign a = ∅
+
+abbrev unAllocated (g : Goods) := ∀ a, g ∉ alloc.assign a
+
+def allocate (g : Goods) (a : Agents) (h : unAllocated alloc g): @Alloc Agents Goods _ := {
+  assign := fun (x : Agents) => if x = a then (alloc.assign x) ∪ {g} else alloc.assign x
+  disjoint := by
+    intro arb_a arb_b hneq
+    simp
+    by_cases ha : arb_a = a
+    case pos =>
+      have hneqb : arb_b ≠ a := by
+        by_contra heqarb_a
+        simp_all only [ne_eq, not_true_eq_false]
+      simp [ha, hneqb, Finset.union_inter_distrib_right]
+      rw [←ha, alloc.disjoint]
+      simp [h]
+      exact hneq
+    case neg =>
+      by_cases hb : arb_b = a <;>
+        simp[hb, ha, alloc.disjoint, hneq]
+      rw[←hb]
+      simp[Finset.inter_union_distrib_left, hneq, alloc.disjoint]
+      simp [h]
+}
+
+def isEmptyAlloc  :=
+  ∀ a : Agents, alloc.assign a = ∅
```

Comment:
This is the same as `alloc.assign = ⊥`, which saves you a definition. You could quite reasonably even implement a `Bot Alloc` instance to make this `alloc = ⊥`.